ext/ddtrace_profiling_native_extension/collectors_idle_sampling_helper.c

```diff
@@ -0,0 +1,241 @@
+
+  reset_state(state);
+
+  return Qtrue;
+}
+
+static VALUE _native_idle_sampling_loop(DDTRACE_UNUSED VALUE self, VALUE self_instance) {
+  struct idle_sampling_loop_state *state;
+  TypedData_Get_Struct(self_instance, struct idle_sampling_loop_state, &idle_sampling_helper_typed_data, state);
+
+  // Release GVL and run the loop waiting for requests
+  rb_thread_call_without_gvl(run_idle_sampling_loop, state, interrupt_idle_sampling_loop, state);
+
+  return Qtrue;
+}
+
+static void *run_idle_sampling_loop(void *state_ptr) {
+  struct idle_sampling_loop_state *state = (struct idle_sampling_loop_state *) state_ptr;
+  int error = 0;
+
+  while (true) {
+    ENFORCE_SUCCESS_NO_GVL(pthread_mutex_lock(&state->wakeup_mutex));
+
+    action next_action;
+    void (*run_action_function)(void);
+
+    // Await for an action
+    while ((next_action = state->requested_action) == ACTION_WAIT) {
+      error = pthread_cond_wait(&state->wakeup, &state->wakeup_mutex);
+      if (error) {
+        // If something went wrong, try to leave the mutex unlocked at least
+        pthread_mutex_unlock(&state->wakeup_mutex);
+        ENFORCE_SUCCESS_NO_GVL(error);
+      }
+    }
+
+    // There's an action to be taken!
+
+    // Record function, if any
+    run_action_function = state->run_action_function;
+
+    // Reset buffer for next request
+    state->requested_action = ACTION_WAIT;
+
+    // Unlock the mutex immediately so other threads can continue to request actions without blocking
+    ENFORCE_SUCCESS_NO_GVL(pthread_mutex_unlock(&state->wakeup_mutex));
+
+    // Process pending action
+    if (next_action == ACTION_RUN) {
+      if (run_action_function == NULL) {
+        grab_gvl_and_raise(rb_eRuntimeError, "Unexpected NULL run_action_function in run_idle_sampling_loop");
+      }
+
+      run_action_function();
+    } else { // ACTION_STOP
+      return NULL;
+    }
+  }
+}
+
+static void interrupt_idle_sampling_loop(void *state_ptr) {
+  struct idle_sampling_loop_state *state = (struct idle_sampling_loop_state *) state_ptr;
+  int error = 0;
+
+  // Note about the error handling in this situation: Something bad happening at this stage is really really awkward to
+  // handle because we get called by the VM in a situation where we can't really raise exceptions, and the VM really really
+  // just wants us to stop what we're doing and return control of the thread to it.
+  //
+  // So if we return immediately on error, we may leave the VM hanging because we didn't actually interrupt the thread.
+  // We're also not at a great location to flag errors.
+  // That's why: a) I chose to log to stderr, as a last-ditch effort; b) even if something goes wrong we still try to
+  // ask the thread to stop, instead of exiting early.
+
+  error = pthread_mutex_lock(&state->wakeup_mutex);
+  if (error) { fprintf(stderr, "[DDTRACE] Error during pthread_mutex_lock in interrupt_idle_sampling_loop (%s)\n", strerror(error)); }
```

Comment:
Are you using this `fprintf(stderr` error pattern frequently?
If you foresee this usage in the future, we can add another error checking macro.

Reply:
Nah, it's very specific to this case.

Reply:
Yup, I really don't want to make it a common thing, was a bit of a last-ditch effort here.

If it does become a common thing, I definitely will extract it and perhaps even make it slightly more configurable; this case was really the "I think it's important to flag when this happens, but there's no good way right now".